go/pkg/cas/upload.go

```diff
@@ -788,7 +790,10 @@ chunkLoop:
 		req.Data = buf[:n] // must limit by `:n` in ErrUnexpectedEOF case
 
 		// Send the chunk.
-		switch err = stream.Send(req); {
+		withTimeout(func() {
+			err = stream.Send(req)
```

Comment:
IIUC, even if this "times out", we'd still have to wait until the hanging request returns so we can stop streaming at the next loop iteration. 

Do you think we can handle this in an more aggressive way? Eg launch the `Send` in a separate goroutine, and moving on to an early return if the time ticks early. I _think_ that even though it doesn't seem like we can forcefully close the stream, letting its resources leak while early returning the error to the user provides a better UX. (FWIW, I'm not sure how complicated the implementation would end up ending up like).

Reply:
we do have to wait for `Send()` to return, but I think it will return quickly because `stream`'s context is canceled. I found a couple of discussions about this: https://github.com/grpc/grpc-go/issues/990 and https://github.com/grpc/grpc-go/issues/3230#issuecomment-562490274 that seem to suggest that `Send()` would be unblocked. FWIU, once the context is canceled, there is not much `StreamClient` can do, other than fail, which is what we are trying to achieve.

Also I am generally worried about leaks :)

Reply:
Ah, good to know that cancelling the ctx does close the stream!